flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/HITSAlgorithm.java

```diff
@@ -0,0 +1,183 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.graph.library;
+
+import org.apache.flink.api.common.aggregators.DoubleSumAggregator;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.java.DataSet;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.EdgeDirection;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.GraphAlgorithm;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.graph.spargel.MessageIterator;
+import org.apache.flink.graph.spargel.MessagingFunction;
+import org.apache.flink.graph.spargel.ScatterGatherConfiguration;
+import org.apache.flink.graph.spargel.VertexUpdateFunction;
+import org.apache.flink.graph.utils.NullValueEdgeMapper;
+import org.apache.flink.types.DoubleValue;
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * This is an implementation of HITS algorithm, using a scatter-gather iteration.
+ * The user can define the maximum number of iterations. HITS algorithm is determined by two parameters,
+ * hubs and authorities. A good hub represents a page that points to many other pages, and a good authority
+ * represented a page that is linked by many different hubs.
+ * Each vertex has a value of Tuple2 type, the first field is hub score and the second field is authority score.
+ * The implementation assumes that the two score are the same in each vertex at the beginning.
+ * <p>
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/HITS_algorithm">HITS Algorithm</a>
+ */
+public class HITSAlgorithm<K, VV, EV> implements GraphAlgorithm<K, VV, EV, DataSet<Vertex<K, Tuple2<Double, Double>>>> {
+
+	private int maxIterations;
+
+	/**
+	 * Creates an instance of HITS algorithm.
+	 *
+	 * @param maxIterations the maximum number of iterations
+	 */
+	public HITSAlgorithm(int maxIterations) {
+		Preconditions.checkArgument(maxIterations > 0, "The number of maximum iteration should be greater than 0.");
+		this.maxIterations = maxIterations * 2 + 1;
+	}
+
+	@Override
+	public DataSet<Vertex<K, Tuple2<Double, Double>>> run(Graph<K, VV, EV> netGraph) throws Exception {
+
+		ScatterGatherConfiguration parameter = new ScatterGatherConfiguration();
+		parameter.setDirection(EdgeDirection.ALL);
+		parameter.registerAggregator("sumVertexValue", new DoubleSumAggregator());
+
+		return netGraph
+				.mapVertices(new VertexInitMapper<K, VV>())
+				.mapEdges(new NullValueEdgeMapper<K, EV>())
```

Comment:
Mapping to `NullValue` is not necessary if the edge values are already `NullValue`. Based on the code in `Translate`, we can `TypeInformation<EV> typeInfo = ((TupleTypeInfo<Edge<K, EV>>) netGraph.getEdges().getType()).getTypeAt(2);` and then `if (typeInfo.getTypeClass().equals(NullValue.class)) ...`